File: contrib/bridge_sampling.R
# run bridge sampling with Stan MCMC chains on Stan models

dir.create(Sys.getenv("R_LIBS_USER"), recursive = TRUE)
.libPaths(Sys.getenv("R_LIBS_USER"))

if (!require("pacman")) install.packages("pacman")
pacman::p_load("bridgesampling", "rstan", "jsonlite")


bridge_sampling = function(target, iter, warmup, chains, seed, headers) {
  set.seed(seed)
  rstan_options(auto_write = TRUE)
  wd <- getwd()
  setwd(headers)

  stanfit <- stan(file = paste(target, ".stan", sep = ""),
                  data = fromJSON(paste(target, ".data.json", sep = "")),
                  iter = iter,
                  warmup = warmup,
                  chains = chains,
                  seed = seed)
        
  setwd(wd)
        
  sampler_params <- get_sampler_params(stanfit, inc_warmup = TRUE)
  mcmc_neval <- sum(do.call(rbind, sampler_params)[,'n_leapfrog__'])

  result <- bridge_sampler(stanfit)
  bs_neval <- length(result$q11) + length(result$q21)
  neval <- mcmc_neval + bs_neval
  logz <- result$logml
  err_logz <- sqrt(error_measures(result)$re2)
  return(c(logz, err_logz, neval))
}

File: contrib/bridge_sampling.py
"""
Run bridge sampling
===================
"""

import os
import rpy2.robjects

CWD = os.path.dirname(os.path.realpath(__file__))
SCRIPT = os.path.join(CWD, "bridge_sampling.R")
HEADERS = os.path.normpath(os.path.join(CWD, "..", "stanfunctions"))

R_OBJECT = rpy2.robjects.r
R_OBJECT.source(SCRIPT)
BRIDGE_SAMPLING = R_OBJECT.bridge_sampling


def run_bridge_sampling(stan_file, iter_=20_000, warmup=5_000, chains=4, seed=0):
    target = os.path.splitext(stan_file)[0]
    data = list(BRIDGE_SAMPLING(
        target, iter_, warmup, chains, seed, HEADERS))
    data[-1] = int(data[-1])
    keys = ["log evidence", "error log evidence", "neval"]
    return dict(zip(keys, data))
